source/utilities.cc

```diff
@@ -1847,6 +1847,163 @@ namespace aspect
+      double averaged_parameter = 0.0;
+
+      // first look at the specal case which can be done faster
+      if (p <= -1000)
+        {
+          // Minimum
+          const unsigned int i = (unsigned int)(std::min_element( weights.begin(),
+                                                                  weights.end() )
+                                                - weights.begin());
+          return values[i];
+        }
+      else if (p == -1)
+        {
+          // Harmonic average (TODO: Check if we need to use a epsilon)
+          for (unsigned int i=0; i< weights.size(); ++i)
+            {
+              if (values[i] == 0 && weights[i] != 0)
+                return 0;
+              else if (values[i] !=0)
+                averaged_parameter += weights[i]/(values[i]);
+            }
+          const double sum_of_weights = std::accumulate(weights.rbegin(), weights.rend(), 0);
+          return sum_of_weights/averaged_parameter;
+        }
+      else if (p == 0)
+        {
+          // Geometric average (TODO: Check if we need to use a epsilon)
+          for (unsigned int i=0; i < weights.size(); ++i)
+            averaged_parameter += weights[i]*std::log(values[i]);
+          const double sum_of_weights = std::accumulate(weights.rbegin(), weights.rend(), 0);
+          return std::exp(averaged_parameter/sum_of_weights);
+        }
+      else if (p == 1)
+        {
+          // Arithmetic average (TODO: Check if we need to use a epsilon)
+          for (unsigned int i=0; i< weights.size(); ++i)
+            averaged_parameter += weights[i]*values[i];
+          const double sum_of_weights = std::accumulate(weights.rbegin(), weights.rend(), 0);
+          return averaged_parameter/sum_of_weights;
+        }
+      else if (p == 2)
+        {
+          // Quadratic average (RMS) (TODO: Check if we need to use a epsilon)
+          for (unsigned int i=0; i< weights.size(); ++i)
+            averaged_parameter += weights[i]*values[i]*values[i];
+          const double sum_of_weights = std::accumulate(weights.rbegin(), weights.rend(), 0);
+          return std::sqrt(averaged_parameter/sum_of_weights);
+        }
+      else if (p == 3)
+        {
+          // Cubic average (TODO: Check if we need to use a epsilon)
+          for (unsigned int i=0; i< weights.size(); ++i)
+            averaged_parameter += weights[i]*values[i]*values[i]*values[i];
+          const double sum_of_weights = std::accumulate(weights.rbegin(), weights.rend(), 0);
+          return std::cbrt(averaged_parameter/sum_of_weights);
+        }
+      else if (p >= 1000)
+        {
+          // Maximum
+          const unsigned int i = (unsigned int)(std::max_element( weights.begin(),
+                                                                  weights.end() )
+                                                - weights.begin());
+          return values[i];
+        }
+
+      for (unsigned int i=0; i< weights.size(); ++i)
```

Comment:
enclose this block into an `else` to make it clear that this is the last fall-through of the sequence of `if`-`else` statements